CppCoreGuidelines.md

```diff
@@ -19182,26 +19182,25 @@ A header should include all its dependencies. Be careful about using relative pa
 
 A test should verify that the header file itself compiles or that a cpp file which only includes the header file compiles.
 
-### <a name="Rs-incform"></a>SF.12: Prefer the angle bracket form of `#include` where you can and the quoted form everywhere else
+### <a name="Rs-incform"></a>SF.12: Prefer the quoted form of `#include` for files relative to the including file and the angle bracket form everywhere else
 
 ##### Reason
 
 The [standard](http://eel.is/c++draft/cpp.include) provides flexibility for compilers to implement
 the two forms of `#include` selected using the angle (`<>`) or quoted (`""`) syntax. Vendors take
 advantage of this and use different search algorithms and methods for specifying the include path.
 
-Nevertheless, the guidance is to use the angle form when possible. This supports the fact that the
-standard library headers must be included this way, is more likely to create portable code, and enables
-the quoted form for other uses. For example being clear about the locality of the header relative
-to files that includes it or in scenarios where the different search algorithm is required.
+Nevertheless, the guidance is to use the quoted form for including files that exist at a relative path to the file containing the `#include` statement and to use the angle bracket form everywhere else, where possible. This encourages being clear about the locality of the header relative to files that include it, or scenarios where the different search algorithm is required. For example, it makes it easy to understand at a glance whether a header is being included from a local relative file versus a standard library header or an external header from another project.
```

Comment:
While much improved, there's still ambiguity in this paragraph. Consider a library project, _proj_, consisting of a source file, _/src/abc.cpp_ and a header file, _/include/proj/abc.h_. The header search path provided to the compiler should be _/include_ and the header should be included using `#include <proj/abc.h>` -- not only in source files from "another project" which depends on _proj_ but also -- in _abs.cpp_.

Because `<` and `>` are used in this example, it's impossible "to understand at a glance whether a header is being included from a local relative file versus a standard library header or an external header from another project."

In general, I think it's just not helpful to introduce the notion of a project into the reasoning.

Reply:
the examples are just some typical cases, to be more precise it could be worded as "or a header from the _alternate search path_". I'll try shifting the 'for example" closer to the project to make it clear it's not an exhaustive set:

"It makes it easy to understand at a glance whether a header is being included from a local relative file versus a standard library header or a header from the alternate search path (for example, an external project header)."

Reply:
Much better, thanks! And thank you for the PR. I've ranted about this issue on [multiple](https://github.com/isocpp/CppCoreGuidelines/pull/1456#issuecomment-564299160) [occasions](https://github.com/isocpp/CppCoreGuidelines/pull/1570#issuecomment-596267124) but never gotten 'round to proposing a fix!

Reply:
> It makes it easy to understand at a glance whether a header is being included from a local relative file versus a standard library header or a header from the alternate search path (for example, an external project header).

I actually agree with this wording: if someone is trying to include a header via "local relative file [path]," I definitely do want them to use `""` instead of `<>`.  My only bone to pick is that I _never want them to try to include a header via local relative file!_ In other words: don't think of the `<>` search path as the "alternate search path"; think of it as _the_ search path. If you want compilation to pick up a header file, then you should make sure that that header file's parent directory is part of _the_ search path.

The bad situation I want to avoid is that you have two different search paths simultaneously in the same compilation process, so that `#include "foo.h"` and `#include <foo.h>` both appear in the codebase, but have different effects. Or, almost as bad: their effects are identical, and yet you have some components that do `#include "foo.h"` and other components that do `#include <foo.h>`, essentially at random, based on whether the original coder considered foo.h to be "part of their [team's] project" or not.

Reply:
the wording change John meant in was to make it more clear that <> isn't just for external projects, but it can also be used within the same project. So I adjusted it to make it clear that external library usage was just one typical example.

Quux, your comment deserves its own thread, and it speaks to the original intent of the PR to address "devolving all usage into <>"; which isn't necessarily a bad thing! C++ has given us this weird directive with two forms. One approach is to collapse the forms into one or we can try to harness a meaningful distinction. I think since modern compilers recognize that "" form searches locally first, we have an opportunity to control compiler behavior directly in code (which is easier to have guidelines for versus compiler options).
#include "utils/foo.h" encodes intent directly into the code that the author wants foo.h from nearby, and not say, foo.h from external project "utils". Using only a single include form (<>) requires specifying this intent in compiler options, which is a more difficult subject to broach here.

Reply:
> I think since modern compilers recognize that "" form searches locally first, we have an opportunity to control compiler behavior directly in code (which is easier to have guidelines for versus compiler options). [...] Using only a single include form (<>) requires specifying this intent in compiler options, which is a more difficult subject to broach here.

Sure, but I'd like the Guidelines if possible to reflect _good coding style_ first and foremost. Reflecting _the aversion C++ programmers have to talking about build systems_ is not a goal for me. Maybe you meant something else. Anyway, every programmer will have to learn how to set their include path sooner or later. The only question is whether they're learning how to set two independent include paths (for `""` and `<>`) or just one.

Reply:
Yes, programmers will still need to know compiler options (in the least for adjusting the set of <> includes), but if the guideline makes the intent clear from code, it won't be ambiguous how to set those options. Right now it's not clear what you need to do to your compiler to compile #include "utils/foo.h". With this PR to SF.12 the answer is collapsed from 4 different possibilities: you don't need to do _anything_ because the default compiler treatment for "" is the correct one. This removes the variability and let's programmers across different environments and compilers to understand from the code how their compiler settings should be adjusted.

Reply:
@Quuxplusone "Reflecting the aversion C++ programmers have to talking about build systems is not a goal for" any of us. Use of the term project here is simply unhelpful because some projects are large enough they they have multiple header search paths of their own, making them indistinguishable from multiple projects for the sake of the arguments subsequently put foward.

And that's the crux. You keep talking about 'the search path' as if `-I` or `/I` is only ever passed to the compiler once. A typical compiler command line is looooong and mentions at least one header search path for _each_ dependency the source needs. *The order of these flags affects the order in which GCC/MSVC searches those paths for files.* That in turn affects the code the preprocessor generates.

However, when using `""`, the relative path gets searched first. Thus `""` is incredibly valuable in bringing stability to the process. Again, please read the [SO link](https://stackoverflow.com/a/52144130/671509) I shared with you for an example.